proc_dash/plotting.py

```diff
@@ -0,0 +1,91 @@
+import pandas as pd
+import plotly.express as px
+
+STATUS_CMAP = px.colors.qualitative.Bold
+STATUS_COLORS = {
+    "SUCCESS": STATUS_CMAP[5],
+    "FAIL": STATUS_CMAP[9],
+    "INCOMPLETE": STATUS_CMAP[3],
+    "UNAVAILABLE": STATUS_CMAP[10],
+}
+PIPELINE_STATUS_ORDER = ["SUCCESS", "FAIL", "INCOMPLETE", "UNAVAILABLE"]
+LAYOUTS = {
+    "margin": {"l": 30, "r": 30, "t": 60, "b": 30},  # margins of chart
+    "title": {  # figure title position properties
+        "yref": "container",
+        "y": 1,
+        "yanchor": "top",
+        "pad": {"t": 20},
+    },
+}
+
+
+def plot_pipeline_status_by_participants(data: pd.DataFrame):
+    long_data = pd.melt(
+        data,
+        id_vars=["participant_id", "session"],
+        var_name="pipeline_name",
+        value_name="pipeline_complete",
+    )
+    status_counts = (
+        long_data.groupby(["pipeline_name", "pipeline_complete", "session"])
+        .size()
+        .reset_index(name="participants")
+    )
```

Comment:
Hrm, do you think this is necessary? A similar but not identical version of this command is used by my two current plotting functions:
# plot_pipeline_by_participants
...
    status_counts = (
        long_data.groupby(["pipeline_name", "pipeline_complete", "session"])
        .size()
        .reset_index(name="participants")
    )
...


# function 2
...
    status_counts = (
        long_data.groupby(["pipeline_name", "pipeline_complete"])
        .size()
        .reset_index(name="records")
    )
...
which you suggested to refactor into its own function. I started to create a get_status_counts  function that handles this command, but now I'm wondering if it's actually necessary - since the groupby variables and reset_index name are different across these two, it feels like I'd just be rewriting this line essentially to call the refactored function

Reply:
Hrm, do you think this is necessary? Since a similar but not identical version of this command is used by my two current plotting functions:

```python
# plot_pipeline_by_participants
...
status_counts = (
long_data.groupby(["pipeline_name", "pipeline_complete", "session"])
.size()
.reset_index(name="participants")
)
...

# plot_pipeline_by_records
...
status_counts = (
long_data.groupby(["pipeline_name", "pipeline_complete"])
.size()
.reset_index(name="records")
)
...
```
(both the `groupby` and `reset_index` params are different), it feels like I'd just be rewriting this line essentially to call a refactored function with appropriate inputs.

Reply:
Ah yes, that's true. Probably best to not refactor then, as you said.